### ISSUE TITLE
SQLite backed cache: smarter WebSockets updates for paginated resources

### ISSUE THREAD

**Issue by gaktive** [rancher/rancher#40773]
Parent backend epic: https://github.com/rancher/rancher/issues/46348 (iteration 2)
Parent backend epic: https://github.com/rancher/rancher/issues/48042 (iteration 3)
Parent frontent epic: rancher/dashboard#5243

Internal reference [SURE-5734](https://jira.suse.com/browse/SURE-5734)
Originally requested in SURE-5698 (2022-12-15)

**Problem**: the auto-refreshing mechanism of paginated resources over WebSockets is currently broken when the SQLite backend cache is enabled. That is because the cache works in a substantially different way and Steve code was not updated to reflect that.

This issue and the accompanying [dashboard](rancher/dashboard#5243) issue are about implementing this mechanism from scratch, with a new design that into account the potential (and downsides) of the new cache.

Scope of this issue is specifically about the backend/Steve parts of the change.

**Proposed solution (overall idea)**:

Steve notifies the UI whenever resource change - no matter if they are currently displayed or not (Steve does not know).

Steve debounces notifications - they are never too frequent.

UI can, depending on UX considerations, react to a notification by `list`ing the currently displayed page of resources again. This will cause more frequent `list` calls, but is expected to be much cheaper as only the relevant page is served, and Kubernetes API is not exercised.

**Proposed solution (Steve perspective)**:
 - Steve gets notified of resource changes via the SQL cache's existing Informer (instead of opening an ad-hoc Kubernetes `Watch` call as it was before). Upon receiving a change event:
   - it increments a per-resource-type `resourceVersions` map of counters. Each entry maps a type with the count of events received in total
   - it checks if the changed resource is visible to the current user. If so:
   - it sends the UI a notification that a change has happened (eg. `{"name":"resources.changed", "resourceType":"management.cattle.io.cluster"}`)
        - Note: such notification does not include any information about the change itself (diff), or the changed resource (eg. id)
        - Note: Steve does not know whether the changed resource is displayed or not
        - Note: Steve implements a configurable "debouncing" mechanism to avoid sending the UI change events too frequently
 - when Steve returns results for any `list` call, it includes the current value of `resourceVersion` for the listed type
 - Steve keeps the current interface for the UI to subscribe to changes intact (eg. `{resourceType: "provisioning.cattle.io.cluster", resourceVersion: "1234"}`)
   -  Note: `resourceVersion` is a value UI code takes from the output of the latest `list` call on that resource type
   -  Note: if `resourceVersion` is missing, the latest version is assumed
   -  Note: if the specified `resourceVersion` is older than the current counter in `resourceVersions`, a change notification is immediately dispatched upon subscription

**Expected outcome (UI perspective)**:
 - UI subscribes to changes to all resources of a certain type, as visible to a certain user, as it happens in current code (eg. `{resourceType: "provisioning.cattle.io.cluster", resourceVersion: "1234"}`)
   - Note: `410 Gone` is not returned any more
 - `resourceVersion` is returned from `list` calls like in current code
 - Steve notifies the UI, via WebSocket, whenever any subscribed resources change
 	- Steve does not include any information about the changes themselves (eg. the changed object). The notification message is much shorter than before (eg. `{"name":"resources.changed","resourceType":"management.cattle.io.cluster"}`)
 	- Steve implements a configurable "debouncing" mechanism to avoid sending the UI change events too frequently. Notifications will not be as frequent as before
 - Upon receiving a change event via WebSocket, the UI has the option to send a new `list` request to re-fetch the visible resource page. This is expected to be much faster than before in most cases (only retrieving a page), therefore it is expected to be performant enough
 - UI may opt not issue a `list` request in response to a change event, for example because autorefresh is turned off for a certain type or by a certain user, or because a notification "Information on this page is stale. Click Refresh" is deemed sufficient

_(2024-08-09: description reworked after discussion downthread)_

**Comment by richard-cox** [comment]
The description contains a lot of good information, but i just want to clarify

- Currently watching happens on all resources of a type. This can lead to a LOT of resource.[created|changed|removed) socket updates which impacts browser performance due to changing the local cache (causing change detection churn) and re-paginating excessively large lists locally 
- In the new world where we use steve pagination to retrieve a single pages worth of results we're seeking a way to stop the flood of resource updates over socket

I've chatted to Neil and Ken and would like to request the following solution

- The UI requests a paginated list
- It receives a single page with a revision. 
- The UI subscribes for pagination changed web socket messages. This contains
  - A revision - This indicates the point in time the original API request was made
  - A page (or all general pagination settings). 
- The BE sends pagination changed web socket message since and following the revision time
  - A pagination changed message indicates that something has changed the target page
- The UI receives a pagination changed message and refetches the current page
  - For example
    - a resource state changes such that it now appears on the current page
    - a new resource is added on an earlier page, changing the last entry on the current page
  - We might be able to abstract this to just
    - a resource is added / removed
    - a resource field has changed that is defined in the sort or filter properties
  - Note - current functionality is adhered to (if requested page exceeds total pages the last page is returned)

Slightly Related
- The above is the only place where the UI might use a revision. The following events will request the latest results
  - Changing page
  - Changing sort order
  - Changing filtering

**Comment by moio** [comment]
I think the above asks are reasonable and make sense, thanks for putting the list together.

In my mind, most of the work on this issue cannot start until work on the caching mechanisms is completed or anyway more advanced than it is now - but it's good to have this indication so that it's kept in mind along the way.

**Comment by richard-cox** [comment]
Building / Further Clarrifying https://github.com/rancher/rancher/issues/40773#issuecomment-1777149286. Needs discussion (contains functionality... but might be spammy)

UI makes a new `pagination` watch request
- contains
  - resource revision taken from the last paginated API request 
  - a collection of resource id's in the current page

Backend sends a "Paginated Web Socket Message - Page Changed"
- Triggered when
  - a resource of this type has been added / removed 
  - a resource's field has changed which is classed a sortable/filterable

Backend sends a Paginated Web Socket Message - Resource in current page changed"
- Triggered when 
  - a resource specified in the watch event changes
  - Note - if it's a field that is sortable/filterable then only `page changed` event is sent
- contains
  - the latest value of the resources specified in the watch event

**Comment by moio** [comment]
> Browser --> Rancher Websocket Event - UI makes a new pagination watch request
> 
>    contains
>        resource revision taken from the last paginated API request
>        a collection of resource id's in the current page

I am not sure this can get what we want.

Suppose I am on page 2 of 10 and suddenly all items in page 1 disappear. At the end of all updates what the user wants to see is (FMPOV) either:
 - I am now moved to page 1, and the same rows as before are displayed, or
 - I am still on page 2, and all rows are now swapped with the rows of former page 3

With the specified data, backend has no possibility to return the right response to the UI (respectively: "please remove page 1 in the paginator widget" or "please swap the displayed rows with these (XXX)").

The only chance would be for the watch call to have all the filtering, sorting and pagination parameters of the original "list" call, and for the backend watch implementation to recompute, for each event, whether:

 - the event (addition/removal/change) affects a row "later than" the displayed page. In that case:
    - change events have no visible effect and can be discarded
    - addition events may add pages
    - deletion events may delete pages
  - the event affects a row in the displayed page. In that case:
    - change events should lead to the displayed row being updated
    - addition events should add the new row in the displayed page, kick the very last row out of the displayed list, and potentially add a page
    - deletion events may bring in an existing row from the next page (note this new row is not the one the event is about(*)!) and potentially delete a page
  - the event affects a row "earlier than" the displayed page. In that case:
    - change events have no visible effect and can be discarded
    - addition events should add a row in the displayed list (note: this new row is not the one the event is about(*)!), kick the very last row out of the displayed list, and potentially add a page
    - deletion events will kick out the very first row in the page, may bring in an existing row from the next page (note this new row is not the one the event is about(*)!) and potentially delete a page


The above all sounds pretty complicated to me - especially because the backend would always need to recompute the position of the event with respect of the displayed page and in some cases (*) also fetch a row from the data set that has nothing to do with the event itself. This is likely to be close in performance cost to redo the list from scratch.


How about a radically simpler approach, where:
 - any changes to the underlying data set are just notified to the UI and the UI requests a list from scratch
 - the maximum change rate is capped ("debounce mechanism") - something that is needed anyway to protect in case of very frequent updates

**Comment by richard-cox** [comment]
> Suppose I am on page 2 of 10 and suddenly all items in page 1 disappear. At the end of all updates what the user wants to see is (FMPOV) either:
> 
> - I am now moved to page 1, and the same rows as before are displayed, or
> - I am still on page 2, and all rows are now swapped with the rows of former page 3
>
>  With the specified data, backend has no possibility to return the right response to the UI (respectively: "please remove page 1 in > the paginator widget" or "please swap the displayed rows with these (XXX)").

I think there was some confusion over my suggestion. The two messages in https://github.com/rancher/rancher/issues/40773#issuecomment-2247977889 cover 

1) A change has occurred that _could_ affect the rows within the provided page. 
   - This is basically has _any_ resource been created / deleted, or has a property in the resource change that is used in sorting / filtering. 
   - Both conditions are regardless of if the resources is in the provided list
   - Examples
     - resource added to previous page
     - resource user is filtering on name, resource on later page name is changed
     - resource is sorting by state, resource on current page state changes
2) If not covered by above, and a change has occurred that does not affect the page just the existing rows it contains.
   - A resource in the provided list of resources has a property that has changed that is not used in sorting / filtering
   - These changes can be sent directly over socket
   - Examples
     - resource annotation is updated, table does nothing with annotaion
     - certificate in a `secret` is updated
     - cluster condition is updated (these could churn a lot and info is just displayed)

Benefits
- Primarily reducing the amount of times the ui may need to request a new page. Specifically when a resource is updated that will not change pages
  - it is not a consideration of sorting or filtering 
  - it is outside of the current page
  - it is inside the current page (the updates are sent with the socket message)
Negatives
- Everytime the UI requests a new page (user filters, changes page, sort order, etc) the UI would need to re-supply the page
- There's more complexity in the backend, plus having to track individual user's page context

The other option is as you suggest, the UI just recieves a notification when there's any change to a resource type (new, updated, removed). UI side we could try and do some smarts to avoid making new page requests, though we would still need some indication on what's changed

**Comment by moio** [comment]
@richard-cox I like your analysis here.

So Frontend asks for something like `resource/list?filter=AAA&sort=BBB`, gets a list of resources with their IDs

Then Frontend asks for `resource/watch?filter=AAA&sort=BBB&resourceRevision=XXX&displayedIDs=1,2,3`

Backend will send events relative to `resource` at or later than XXX. Specifically:
 - additions, deletions: all unconditionally
 - changes: only if either the event's resource ID is in `displayedIDs` or if anything changed on any of the properties present in `filter` or `sort`
 
 Is that a correct understanding?

**Comment by richard-cox** [comment]
Correct, though the triggers would be in two buckets (names very much tbd)
- Page Changed (UI will re-fetch the current page)
  - additions, deletions: all unconditionally
  - resource changes: if anything changed on any of the properties present in filter or sort
- Resource Changed (UI will just update content in a page)
  - resource changes: if the event's resource ID is in displayedIDs and not if anything changed on any of the properties present in filter or sort


There's a few things to work out
- how displayedIds are tracked.
  - For example the user could go through the below flow quite quickly, drastically changing the content of displayedIds. resource/list?sort=BBB --> resource/list?sort=-BBB --> resource/list?sort=-BBB&filter=te --> resource/list?sort=-BBB&filter=test
 - Options
   - ui receives page and then send displayedIds back over socket for the backend to track
   - ui supplies a list id when requesting the resource/list and the backend then caches the page returned against that id).
- Debounced events would pick the event with the highest priority, for example there's no point the ui getting a resource change notification if there's an addition/deletion

I'll admit the proposal to track displayedId's is a pain, open to other options as well

**Comment by moio** [comment]
There's an additional problem with this approach:

> changes: only if either the event's resource ID is in displayedIDs or if anything changed on *any of the properties present in filter or sort**

In case of updates, Kubernetes gives Steve the new object - not the diff between the old and the new.

So in order to distinguish a `Page Changed` from a `Resource Changed` event Steve will need a round-trip to the database - and if it is `Page Changed` then a full page refresh will be needed anyway.

I am not sure this "optimization" is worth the trouble - moreover it will only be correct if `displayedIds` is in sync, which as you pointed out we can already say it cannot possibly be "always". Instead of `displayedIds` we could equivalently give the websocket endpoint the full query params, but that one is also at risk of falling out of sync.

I see two ways out:
 - we drop the optimization attempt, at least until we prove we absolutely need it. Assume fetching a page is fast "enough", treat any event like a `Page Changed` event. At the end of the day, additions + deletions + updates to displayed rows + updates of any row on sensible columns already are `Page Changed` events, which is why I call distinguishing `Resource Changed` events an optimization. Would that be doable from a UI perspective? Do you see adverse effects?

 - we drop the list+watch model completely, transitioning to a "watch exclusively" model, similar to Kubernetes's [Streaming List](https://kubernetes.io/docs/reference/using-api/api-concepts/#streaming-lists). Changing filtering/sorting/page will result in a new message being sent over the WebSocket and a new list returned.
 
To me the second approach is quite a lot more difficult to implement given where we are now, so my concern would be time to deliver. My assumption is it would be prohibitively time-consuming to adapt all UI code as well - but you are better positioned to tell for sure. What do you think?

**Comment by moio** [comment]
> So this would look like...
> 
>     * UI requests a watch on a resource type indicating it just wants updates when a resource of that type is created, updated or deleted
>       * Message content would just be a tweak of the current
>         * `{resourceType: "provisioning.cattle.io.cluster"}`--> TBD `{resourceType: "provisioning.cattle.io.cluster", mode: "summary"}`

+1, or even with no changes to the current message at all - if SQLite caching is enabled, Steve will reply in "summary mode" exclusively (in fact it will not know how to properly reply otherwise).

>     * Steve monitors type and sends debounced updates
>       
>       * maybe the ui could state the debounce time (think this would work given steve -- kube socket is per ui -- steve?). this would allow us to fine tune it, and offer a nicer feature

Yes, the "start watching" message could contain the preferred debounce time (if you believe the UI should control that - on a type-by-type or page-by-page basis), or that could be made a user preference, or configurable installation parameter. The only sure thing is it won't just be hardcoded.

>       * updates would be simplist
>         * `{"name":"resources.changed","resourceType":"management.cattle.io.cluster"}` (note name)

+1

> UX wise i'm wondering then if it we'd have some kind of auto-update feature
>     * Enabled - On receiving message UI automatically requests new page and displays to user
>     * Partial - On receiving message UI notifies user list is stale (probably small icon somewhere). User is free to update page manually
>     * Disabled - UI does not watch for changes. User must update page manually. This is what we currently have
> 
> This UX would mitigate some of the issue of not receiving succint triggers (for "Resource Changed"), i.e. resources that are consistently changing resulting in pages refreshing every time it receives a resources changed message

Some applications do this with a per-component toggle notably Grafana. I think the idea is reasonable, especially where sensible defaults are chosen. Eg. defaulting to no updates + refresh button in cases where updates are not expected to be frequent (eg. users) and autorefreshing in other cases (eg. events).

> > we drop the list+watch model completely, transitioning to a "watch exclusively" model, similar to Kubernetes's [Streaming List](https://kubernetes.io/docs/reference/using-api/api-concepts/#streaming-lists). Changing filtering/sorting/page will result in a new message being sent over the WebSocket and a new list returned.
> 
> Would this mean any changes to the resource would result in the new page automatically sent over socket to the UI, basically chopping out the steve --> ui page changed message, then ui -- steve request for new page?

That's a good question. Actually, thinking better about this problem, I got to the conclusion it would solve any problems, only change their aesthetics a bit.

Fundamentally the idea is about replacing the current `List` (http GET) + `Watch` (WebSocket) pair of requests, with just one "more powerful" `Watch` request. That "powerful `Watch`" would return the same results as `List` via the WebSocket upon opening, and then behaving like `Watch` does today. That has the advantage of not needing a `resourceVersion` "token" to make sure there's no gaps between the requests, as there would be only one request. But other than that, all problems mentioned upthread would remain unchanged - and it's a big change both on the frontend and backend.

As you mentioned I agree that's cleaner, but there are enough ongoing changes by 2.10 already. Let's revisit that discussion in a future version when the SQL cache is well established and stable.

> It both cases it would be interesting to see the frequency of page requests impacts performance.

Yes, looking forward to benchmarks 👍

**Comment by moio** [comment]
@richard-cox I reworded the issue description to capture the current state of the design. Please let me know if anything needs clarification/elaboration/correction!

**Comment by richard-cox** [comment]
I've moved this to P0 given all of the SSP work depends on it

**Comment by tomleb** [comment]
Here's a summary of this task, where we're at, etc. @ericpromislow has been assigned to back me up while I'm in vacation.

# Summary

**You probably do not want to read most of the comments from this issue**. I know.. The issue is old and lots has changed. Alright, now that that's out of the way, here is the real summary..

Neither the UI nor the backend supported SQL-backed watch. We want UI to be able to watch resources (eg: `configmaps`) and the source of data for the watch should be the SQLite database from Steve.

A couple things need to be supported:
- Notification mode: For pagination, we only care about knowing something changed, not what changed. In that context, the UI just send another LIST request to get the latest data. (Think pagination of Deployments, for example).
- Filtering: Pre-vai, we support filtering by name, namespace and label selector when making watches.
- Garbage collection so that we don't store indefinitely events

# List of PRs

I ended up splitting this work into many **many** **MANY** different PRs, so here's a nicer list of them.

First, most of the low-level watch / subscribe functionality is provided by rancher/apiserver:

- https://github.com/rancher/apiserver/pull/160: Events responses from Steve did not include ResourceType and Namespace. That's useful / needed by UI.
-  https://github.com/rancher/apiserver/pull/161: For pagination use case, we don't care about the actual item data, we just want to know when something changed, so that UI can make another LIST request.
- https://github.com/rancher/apiserver/pull/162: For pagination use case, we want to debounce those notifications. If 100 objects are modified, we don't want to receive 100 notifications and then have the browser make 100 LIST requests. So we debounce here and it's configurable by the UI when starting the watch.
- https://github.com/rancher/apiserver/pull/163: A bug in apiserver when watch events are TOO fast which causes the watch to wrongly stop.

Then, on steve side we also have many PRs.

First, some "code organization" bits:
- https://github.com/rancher/steve/pull/643: Remove `limit`, `continue` parameter because we no longer need them. Either the UI wants a page (and it will get the latest value of the chosen page at the time of the request) OR it wants ALL data.
- https://github.com/rancher/steve/pull/644: Some code organization changes. We have this abstraction that imo is slowing us down, I hope we can get rid of it at some point, but for now we're just adding one more case to RegisterX.
- https://github.com/rancher/steve/pull/649: More code organization changes. Same deal here as the previous, except callbacks registered with RegisterAfterDeleteAll is not even used, ever. That's some tech debt we'll have to look at eventually.
- https://github.com/rancher/steve/pull/652: Our tests are full of mocks that don't exercise the code. We know about it, it's tech debt as well, this PR basically tried to fix it for the thing I was working on and it did find some bugs.

Then, actual watch functionality in steve:
- https://github.com/rancher/steve/pull/653: Basic watch support. Nothing is stored yet, we just get event as they come in to the database (but not stored in the _events table)
- https://github.com/rancher/steve/pull/654: The UI needs some way of filtering events. This is unchanged from pre-vai world. Supports very basic filters such as namespace, name and label selectors.
- https://github.com/rancher/steve/pull/661: We're finally storing events in the DB, never removing them.
- https://github.com/rancher/steve/pull/662: Fixing some misleading SQL errors. I say misleading because I lost a whole day debugging this thinking it was something it wasn't.
- https://github.com/rancher/steve/pull/664: We're finally removing events from the DB. It's not the best way of doing it, Peter suggested doing it by batching, but for now that's going to be it. We can revisit when / if it becomes a bottleneck.
- https://github.com/rancher/steve/pull/667: When UI makes a watch request with a resourceversion, if that RV is too old, we need to send an error to the UI to let them know. We were only printing the error, we're not properly sending it back. This wasn't caught because we lack testing at this layer of steve (we only have mock tests). I need to take some time to migrate the tests to be mockless and actually be useful.

Finally, integrating it in Rancher:
- https://github.com/rancher/rancher/pull/50606: Bumping steve and rancher/apiserver and setting a hardcoded (for now) limit of 1000.


# Development

## Code organization (quick links)

The meat of the SQL watch logic is in [steve's pkg/sqlcache/informer/listoption_indexer.go](https://github.com/rancher/steve/blob/cf97607be556a83608b72779a2f8f57bfd88640d/pkg/sqlcache/informer/listoption_indexer.go#L285).

The underlying watch logic is in [rancher/apiserver's pkg/subscribe/watcher.go](https://github.com/rancher/apiserver/blob/9d1ff49767e7e585dbd571223c9386099c475693/pkg/subscribe/watcher.go#L76).

## Testing / running

Here's how you can test this with Steve directly (for faster development if needed for bug fixes).

1. Run Steve with SQL cache enabled in one terminal. (Make sure `KUBECONFIG` env var points to a Kubernetes cluster)

```
go run . --sql-cache --debug
```

2. In another terminal, you can run [websocat](https://github.com/vi/websocat). This is similar to telnet where you'll be able to enter messages (separated by newlines (by pressing enter)).

Here's an example of how to watch configmaps:

```
$ websocat -k wss://localhost:9443/v1/subscribe
{"resourceType":"configmaps"}
```

That's it, the watch is configured and if you create, delete or update namespaces in yet another terminal you should see a json line appear.

Here are a few examples:

**Notification mode**

```
$ websocat -k wss://localhost:9443/v1/subscribe
{"resourceType":"configmaps","mode":"resource.changes"}
```

**Filtering by label selector**

```
$ websocat -k wss://localhost:9443/v1/subscribe
{"resourceType":"configmaps","selector":"app=rancher"}
```

**Starting from a specific resourceVersion**

```
$ websocat -k wss://localhost:9443/v1/subscribe
{"resourceType":"configmaps","resourceVersion":"1234"}
```

## Future work

Note that it was sorta implied that this should work with HA but that proved difficult to do so we split the issue. HA support is not its own issue: https://github.com/rancher/rancher/issues/50603

**Comment by gaktive** [comment]
@ericpromislow will sync with @crobby on this.

**Comment by gaktive** [comment]
This may already be in. @crobby to confirm and pass to QA if so.

**Comment by Jono-SUSE-Rancher** [comment]
Based on conversation from @crobby - it looks like this is ready to test.

**Comment by brudnak** [comment]
<details>
    <summary>🧪 Test Environment... CLICK TO EXPAND! ⬅️</summary>
<br>

| Component                        | Version / Type                      |
| -------------------------------- | ----------------------------------- |
| Rancher version                  | v2.12.0-alpha17                     |
| RKE2 Local K8s Version           | v1.33.0+rke2r1                      |
| Installation option              | Helm (high availability)            |
| Cert Details                     | external tls `aws acm`              |
| Downstream cluster type          | n/a                                 |
| Downstream K8s version           | n/a                                 |
| Authentication providers enabled | local                               |
| Logged in user role              | admin                               |
| Webhook version                  | rancher/rancher-webhook:v0.8.0-rc.8 |

</details>

<a name="top"></a>

### 🧪 Test Cases

| \#  | Priority | Description & Link                          | PASS/FAIL |
| --- | -------- | ------------------------------------------- | --------- |
| 1   | P0       | [Run VAI Enabled Regression Suite](#test-1) | ✅ PASS    |
| 2   | P0       | [VAI Smarter Websockets](#test-2)           | ✅ PASS    |

<details>
    <summary>🚨 2 test cases... CLICK TO EXPAND! (For table links to work) ⬅️</summary>
<br>

# 1 / Run VAI Enabled Regression Suite Status: ✅ PASS

<a name="test-1"></a>

**:small_red_triangle: [back to top](#top)**

<details>
    <summary>Test 1 details... Click to expand</summary>

**Test Steps for Validation**

1. Run vai regression test cases https://github.com/rancher/tests/blob/06cf52b680728bfccb83bff63400448631b5f361/validation/steve/vai/vai_test.go#L980
2. 52 tests passed

```test.log
=== RUN   TestVaiTestSuite
time="2025-07-09T12:50:38-07:00" level=info msg="Feature: ui-sql-cache"
time="2025-07-09T12:50:38-07:00" level=info msg="  spec.value: <nil>"
time="2025-07-09T12:50:38-07:00" level=info msg="  status.default: true"
time="2025-07-09T12:50:38-07:00" level=info msg="  status.description: Improve performance by enabling SQLite-backed caching. This also enables server-side pagination and other scaling based performance improvements."
time="2025-07-09T12:50:38-07:00" level=info msg="  status.dynamic: false"
time="2025-07-09T12:50:38-07:00" level=info msg="  status.lockedValue: <nil>"
time="2025-07-09T12:50:38-07:00" level=info msg="  VAI enabled: true (spec.value is nil, using default)"
=== RUN   TestVaiTestSuite/TestVaiEnabled
=== RUN   TestVaiTestSuite/TestVaiEnabled/SetupTestResources
    vai_test.go:100: Setting up test resources...
    vai_test.go:107: Creating namespace: db-namespace-aaxxmajz
    vai_test.go:115: Creating secret: db-secret-olhojssh
    vai_test.go:130: Accessing metrics endpoint to ensure tables are created
    vai_test.go:137: Listing resources to hydrate VAI cache...
    vai_test.go:142: Waiting 15s to ensure resources are cached...
    vai_test.go:147: Test resources setup completed
--- PASS: TestVaiTestSuite/TestVaiEnabled/SetupTestResources (16.49s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/ExtractAllDatabases
time="2025-07-09T12:50:55-07:00" level=info msg="Extracting VAI databases from all pods..."
time="2025-07-09T12:50:55-07:00" level=info msg="Found 3 Rancher pods"
time="2025-07-09T12:50:55-07:00" level=info msg="Extracting database from pod 1/3: rancher-c7f457747-bxrcl"
time="2025-07-09T12:51:17-07:00" level=info msg="Successfully extracted 3416064 byte SQLite database from pod rancher-c7f457747-bxrcl"
time="2025-07-09T12:51:17-07:00" level=info msg="Successfully extracted database from pod rancher-c7f457747-bxrcl"
time="2025-07-09T12:51:17-07:00" level=info msg="Extracting database from pod 2/3: rancher-c7f457747-gxktq"
time="2025-07-09T12:51:36-07:00" level=info msg="Successfully extracted 3256320 byte SQLite database from pod rancher-c7f457747-gxktq"
time="2025-07-09T12:51:36-07:00" level=info msg="Successfully extracted database from pod rancher-c7f457747-gxktq"
time="2025-07-09T12:51:36-07:00" level=info msg="Extracting database from pod 3/3: rancher-c7f457747-q6ttw"
time="2025-07-09T12:51:55-07:00" level=info msg="Successfully extracted 3342336 byte SQLite database from pod rancher-c7f457747-q6ttw"
time="2025-07-09T12:51:55-07:00" level=info msg="Successfully extracted database from pod rancher-c7f457747-q6ttw"
time="2025-07-09T12:51:55-07:00" level=info msg="Extraction completed. Successfully extracted 3 databases"
--- PASS: TestVaiTestSuite/TestVaiEnabled/ExtractAllDatabases (60.96s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckDBFilesExist
    vai_test.go:157: Checking if VAI database files exist in pods...
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckDBFilesExist/Pod_rancher-c7f457747-gxktq
    vai_test.go:165: Database file for pod rancher-c7f457747-gxktq is valid with 15 tables
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckDBFilesExist/Pod_rancher-c7f457747-gxktq (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckDBFilesExist/Pod_rancher-c7f457747-q6ttw
    vai_test.go:165: Database file for pod rancher-c7f457747-q6ttw is valid with 10 tables
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckDBFilesExist/Pod_rancher-c7f457747-q6ttw (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckDBFilesExist/Pod_rancher-c7f457747-bxrcl
    vai_test.go:165: Database file for pod rancher-c7f457747-bxrcl is valid with 30 tables
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckDBFilesExist/Pod_rancher-c7f457747-bxrcl (0.00s)
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckDBFilesExist (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckSecretInDB
    vai_test.go:171: Checking if secret exists in VAI databases...
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckSecretInDB/Pod_rancher-c7f457747-bxrcl
    vai_test.go:199: Secret db-secret-olhojssh found in pod rancher-c7f457747-bxrcl
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckSecretInDB/Pod_rancher-c7f457747-bxrcl (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckSecretInDB/Pod_rancher-c7f457747-gxktq
    vai_test.go:185: Secret table does not exist
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckSecretInDB/Pod_rancher-c7f457747-gxktq (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckSecretInDB/Pod_rancher-c7f457747-q6ttw
    vai_test.go:185: Secret table does not exist
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckSecretInDB/Pod_rancher-c7f457747-q6ttw (0.00s)
    vai_test.go:207: Secret found in 1 out of 3 pods
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckSecretInDB (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckNamespaceInAllVAIDatabases
    vai_test.go:212: Checking if namespace exists in all VAI databases...
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckNamespaceInAllVAIDatabases/Pod_rancher-c7f457747-bxrcl
    vai_test.go:238: Namespace db-namespace-aaxxmajz found in pod rancher-c7f457747-bxrcl
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckNamespaceInAllVAIDatabases/Pod_rancher-c7f457747-bxrcl (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckNamespaceInAllVAIDatabases/Pod_rancher-c7f457747-gxktq
    vai_test.go:238: Namespace db-namespace-aaxxmajz found in pod rancher-c7f457747-gxktq
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckNamespaceInAllVAIDatabases/Pod_rancher-c7f457747-gxktq (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckNamespaceInAllVAIDatabases/Pod_rancher-c7f457747-q6ttw
    vai_test.go:238: Namespace db-namespace-aaxxmajz found in pod rancher-c7f457747-q6ttw
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckNamespaceInAllVAIDatabases/Pod_rancher-c7f457747-q6ttw (0.00s)
    vai_test.go:246: Namespace found in 3 out of 3 pods
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckNamespaceInAllVAIDatabases (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckMetricTablesInVAIDatabase
    vai_test.go:252: Checking if metric tables exist in VAI databases...
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckMetricTablesInVAIDatabase/Pod_rancher-c7f457747-q6ttw
    vai_test.go:284: No metric tables found (this is acceptable)
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckMetricTablesInVAIDatabase/Pod_rancher-c7f457747-q6ttw (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckMetricTablesInVAIDatabase/Pod_rancher-c7f457747-bxrcl
    vai_test.go:279: Found table metrics.k8s.io_v1beta1_NodeMetrics
    vai_test.go:279: Found table metrics.k8s.io_v1beta1_NodeMetrics_fields
    vai_test.go:279: Found table metrics.k8s.io_v1beta1_NodeMetrics_indices
    vai_test.go:286: All 3 metric tables present
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckMetricTablesInVAIDatabase/Pod_rancher-c7f457747-bxrcl (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckMetricTablesInVAIDatabase/Pod_rancher-c7f457747-gxktq
    vai_test.go:284: No metric tables found (this is acceptable)
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckMetricTablesInVAIDatabase/Pod_rancher-c7f457747-gxktq (0.00s)
    vai_test.go:294: Summary: 1 pods have all metric tables
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests/CheckMetricTablesInVAIDatabase (0.00s)
--- PASS: TestVaiTestSuite/TestVaiEnabled/DatabaseTests (0.00s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretFilters
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretFilters/Filter_with_negation_and_namespace
time="2025-07-09T12:51:55-07:00" level=info msg="Starting case: Filter with negation and namespace"
time="2025-07-09T12:51:55-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:51:55-07:00" level=info msg="Creating namespace: namespace-grhcueup"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretFilters/Filter_with_negation_and_namespace (1.91s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretFilters/Filter_by_namespace
time="2025-07-09T12:51:57-07:00" level=info msg="Starting case: Filter by namespace"
time="2025-07-09T12:51:57-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:51:57-07:00" level=info msg="Creating namespace: namespace1-cboxyftm"
time="2025-07-09T12:51:58-07:00" level=info msg="Creating namespace: namespace2-cboxyftm"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretFilters/Filter_by_namespace (1.41s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretFilters/Filter_by_name_and_namespace
time="2025-07-09T12:51:59-07:00" level=info msg="Starting case: Filter by name and namespace"
time="2025-07-09T12:51:59-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:51:59-07:00" level=info msg="Creating namespace: namespace-mdtpvysg"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretFilters/Filter_by_name_and_namespace (1.35s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretFilters/Filter_by_single_label
time="2025-07-09T12:52:00-07:00" level=info msg="Starting case: Filter by single label"
time="2025-07-09T12:52:00-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:52:00-07:00" level=info msg="Creating namespace: namespace-hwqugliu"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretFilters/Filter_by_single_label (1.14s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretFilters/Filter_by_multiple_labels
time="2025-07-09T12:52:01-07:00" level=info msg="Starting case: Filter by multiple labels"
time="2025-07-09T12:52:01-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:52:01-07:00" level=info msg="Creating namespace: namespace-jhbmklqn"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretFilters/Filter_by_multiple_labels (1.38s)
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretFilters (7.19s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/PodFilters
=== RUN   TestVaiTestSuite/TestVaiEnabled/PodFilters/Filter_by_nginx_image
time="2025-07-09T12:52:03-07:00" level=info msg="Starting case: Filter by nginx image"
time="2025-07-09T12:52:03-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:52:03-07:00" level=info msg="Creating namespace: namespace1-ayyhgeem"
time="2025-07-09T12:52:03-07:00" level=info msg="Creating namespace: namespace2-ayyhgeem"
--- PASS: TestVaiTestSuite/TestVaiEnabled/PodFilters/Filter_by_nginx_image (1.85s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/PodFilters/Filter_by_busybox_image
time="2025-07-09T12:52:05-07:00" level=info msg="Starting case: Filter by busybox image"
time="2025-07-09T12:52:05-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:52:05-07:00" level=info msg="Creating namespace: namespace1-lismvlta"
time="2025-07-09T12:52:05-07:00" level=info msg="Creating namespace: namespace2-lismvlta"
--- PASS: TestVaiTestSuite/TestVaiEnabled/PodFilters/Filter_by_busybox_image (1.95s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/PodFilters/Filter_by_non-existent_image
time="2025-07-09T12:52:06-07:00" level=info msg="Starting case: Filter by non-existent image"
time="2025-07-09T12:52:06-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:52:06-07:00" level=info msg="Creating namespace: namespace1-lwmsjrwr"
time="2025-07-09T12:52:07-07:00" level=info msg="Creating namespace: namespace2-lwmsjrwr"
--- PASS: TestVaiTestSuite/TestVaiEnabled/PodFilters/Filter_by_non-existent_image (1.85s)
--- PASS: TestVaiTestSuite/TestVaiEnabled/PodFilters (5.65s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretSorting
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_name_ascending
time="2025-07-09T12:52:08-07:00" level=info msg="Starting case: Sort by name ascending"
time="2025-07-09T12:52:08-07:00" level=info msg="Running with vai enabled: [true]"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_name_ascending (1.98s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_name_descending
time="2025-07-09T12:52:10-07:00" level=info msg="Starting case: Sort by name descending"
time="2025-07-09T12:52:10-07:00" level=info msg="Running with vai enabled: [true]"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_name_descending (2.08s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_namespace_ascending
time="2025-07-09T12:52:12-07:00" level=info msg="Starting case: Sort by namespace ascending"
time="2025-07-09T12:52:12-07:00" level=info msg="Running with vai enabled: [true]"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_namespace_ascending (2.08s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_namespace_descending
time="2025-07-09T12:52:14-07:00" level=info msg="Starting case: Sort by namespace descending"
time="2025-07-09T12:52:14-07:00" level=info msg="Running with vai enabled: [true]"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_namespace_descending (2.09s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_name_and_namespace_ascending
time="2025-07-09T12:52:17-07:00" level=info msg="Starting case: Sort by name and namespace ascending"
time="2025-07-09T12:52:17-07:00" level=info msg="Running with vai enabled: [true]"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_name_and_namespace_ascending (2.12s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_namespace_and_name_ascending
time="2025-07-09T12:52:19-07:00" level=info msg="Starting case: Sort by namespace and name ascending"
time="2025-07-09T12:52:19-07:00" level=info msg="Running with vai enabled: [true]"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_namespace_and_name_ascending (2.11s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_name_ascending_and_namespace_descending
time="2025-07-09T12:52:21-07:00" level=info msg="Starting case: Sort by name ascending and namespace descending"
time="2025-07-09T12:52:21-07:00" level=info msg="Running with vai enabled: [true]"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_name_ascending_and_namespace_descending (2.08s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_namespace_descending_and_name_ascending
time="2025-07-09T12:52:23-07:00" level=info msg="Starting case: Sort by namespace descending and name ascending"
time="2025-07-09T12:52:23-07:00" level=info msg="Running with vai enabled: [true]"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretSorting/Sort_by_namespace_descending_and_name_ascending (2.11s)
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretSorting (16.66s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/CheckVaiDescription
--- PASS: TestVaiTestSuite/TestVaiEnabled/CheckVaiDescription (0.16s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/TimestampCacheTests
=== RUN   TestVaiTestSuite/TestVaiEnabled/TimestampCacheTests/Pod_Age_field_updates_correctly
time="2025-07-09T12:52:25-07:00" level=info msg="Starting timestamp test: Pod Age field updates correctly"
time="2025-07-09T12:52:25-07:00" level=info msg="Running with VAI enabled: [true]"
time="2025-07-09T12:52:26-07:00" level=info msg="Waiting 5 seconds for initial age..."
time="2025-07-09T12:52:31-07:00" level=info msg="Fields for pod: [test-pod-nsmaaklh 1/1 Running 0 4m53s <REDACTED> ip-<REDACTED> <none> <none>]"
time="2025-07-09T12:52:31-07:00" level=info msg="Check 1 - pod Age: 4m53s (at index 4)"
time="2025-07-09T12:52:31-07:00" level=info msg="Waiting 5s before check 2..."
time="2025-07-09T12:52:36-07:00" level=info msg="Check 2 - pod Age: 4m58s (at index 4)"
time="2025-07-09T12:52:36-07:00" level=info msg="Waiting 5s before check 3..."
time="2025-07-09T12:52:41-07:00" level=info msg="Check 3 - pod Age: 5m8s (at index 4)"
time="2025-07-09T12:52:41-07:00" level=info msg="Verifying Age values are increasing..."
time="2025-07-09T12:52:41-07:00" level=info msg="Total age increase: 15s over 3 checks"
time="2025-07-09T12:52:41-07:00" level=info msg="✓ pod timestamp test passed: Age field correctly updates over time"
--- PASS: TestVaiTestSuite/TestVaiEnabled/TimestampCacheTests/Pod_Age_field_updates_correctly (16.05s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/TimestampCacheTests/Deployment_Age_field_updates_correctly
time="2025-07-09T12:52:41-07:00" level=info msg="Starting timestamp test: Deployment Age field updates correctly"
time="2025-07-09T12:52:41-07:00" level=info msg="Running with VAI enabled: [true]"
time="2025-07-09T12:52:42-07:00" level=info msg="Waiting 5 seconds for initial age..."
time="2025-07-09T12:52:47-07:00" level=info msg="Fields for apps.deployment: [test-deployment-qeejgdsr 1/1 1 1 5m12s test-container nginx:alpine app=test-deployment-qeejgdsr]"
time="2025-07-09T12:52:47-07:00" level=info msg="Check 1 - apps.deployment Age: 5m12s (at index 4)"
time="2025-07-09T12:52:47-07:00" level=info msg="Waiting 5s before check 2..."
time="2025-07-09T12:52:52-07:00" level=info msg="Check 2 - apps.deployment Age: 5m21s (at index 4)"
time="2025-07-09T12:52:52-07:00" level=info msg="Waiting 5s before check 3..."
time="2025-07-09T12:52:57-07:00" level=info msg="Check 3 - apps.deployment Age: 5m26s (at index 4)"
time="2025-07-09T12:52:57-07:00" level=info msg="Verifying Age values are increasing..."
time="2025-07-09T12:52:57-07:00" level=info msg="Total age increase: 14s over 3 checks"
time="2025-07-09T12:52:57-07:00" level=info msg="✓ apps.deployment timestamp test passed: Age field correctly updates over time"
--- PASS: TestVaiTestSuite/TestVaiEnabled/TimestampCacheTests/Deployment_Age_field_updates_correctly (16.18s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/TimestampCacheTests/Service_Age_field_updates_correctly
time="2025-07-09T12:52:57-07:00" level=info msg="Starting timestamp test: Service Age field updates correctly"
time="2025-07-09T12:52:57-07:00" level=info msg="Running with VAI enabled: [true]"
time="2025-07-09T12:52:58-07:00" level=info msg="Waiting 5 seconds for initial age..."
time="2025-07-09T12:53:03-07:00" level=info msg="Fields for service: [test-service-cvditsem ClusterIP 10.43.53.12 <none> 80/TCP 5m30s app=test-service-cvditsem]"
time="2025-07-09T12:53:03-07:00" level=info msg="Check 1 - service Age: 5m30s (at index 5)"
time="2025-07-09T12:53:03-07:00" level=info msg="Waiting 5s before check 2..."
time="2025-07-09T12:53:09-07:00" level=info msg="Check 2 - service Age: 5m35s (at index 5)"
time="2025-07-09T12:53:09-07:00" level=info msg="Waiting 5s before check 3..."
time="2025-07-09T12:53:14-07:00" level=info msg="Check 3 - service Age: 5m40s (at index 5)"
time="2025-07-09T12:53:14-07:00" level=info msg="Verifying Age values are increasing..."
time="2025-07-09T12:53:14-07:00" level=info msg="Total age increase: 10s over 3 checks"
time="2025-07-09T12:53:14-07:00" level=info msg="✓ service timestamp test passed: Age field correctly updates over time"
--- PASS: TestVaiTestSuite/TestVaiEnabled/TimestampCacheTests/Service_Age_field_updates_correctly (16.27s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/TimestampCacheTests/ConfigMap_Age_field_updates_correctly
time="2025-07-09T12:53:14-07:00" level=info msg="Starting timestamp test: ConfigMap Age field updates correctly"
time="2025-07-09T12:53:14-07:00" level=info msg="Running with VAI enabled: [true]"
time="2025-07-09T12:53:14-07:00" level=info msg="Waiting 5 seconds for initial age..."
time="2025-07-09T12:53:20-07:00" level=info msg="Fields for configmap: [test-cm-gioudevj 1 5m43s]"
time="2025-07-09T12:53:20-07:00" level=info msg="Check 1 - configmap Age: 5m43s (at index 2)"
time="2025-07-09T12:53:20-07:00" level=info msg="Waiting 5s before check 2..."
time="2025-07-09T12:53:25-07:00" level=info msg="Check 2 - configmap Age: 5m48s (at index 2)"
time="2025-07-09T12:53:25-07:00" level=info msg="Waiting 5s before check 3..."
time="2025-07-09T12:53:30-07:00" level=info msg="Check 3 - configmap Age: 6m8s (at index 2)"
time="2025-07-09T12:53:30-07:00" level=info msg="Verifying Age values are increasing..."
time="2025-07-09T12:53:30-07:00" level=info msg="Total age increase: 25s over 3 checks"
time="2025-07-09T12:53:30-07:00" level=info msg="✓ configmap timestamp test passed: Age field correctly updates over time"
--- PASS: TestVaiTestSuite/TestVaiEnabled/TimestampCacheTests/ConfigMap_Age_field_updates_correctly (16.42s)
--- PASS: TestVaiTestSuite/TestVaiEnabled/TimestampCacheTests (64.92s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretPageSize
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretPageSize/Paginate_50_secrets_with_pagesize_10
time="2025-07-09T12:53:30-07:00" level=info msg="Starting case: Paginate 50 secrets with pagesize 10"
time="2025-07-09T12:53:30-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:53:57-07:00" level=info msg="First page retrieved 10 items"
time="2025-07-09T12:53:58-07:00" level=info msg="✓ Retrieved 50 secrets across 5 pages (expected 5 pages with pagesize 10)"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretPageSize/Paginate_50_secrets_with_pagesize_10 (28.30s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretPageSize/Paginate_100_secrets_with_pagesize_25
time="2025-07-09T12:53:58-07:00" level=info msg="Starting case: Paginate 100 secrets with pagesize 25"
time="2025-07-09T12:53:58-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:54:42-07:00" level=info msg="First page retrieved 25 items"
time="2025-07-09T12:54:42-07:00" level=info msg="✓ Retrieved 100 secrets across 4 pages (expected 4 pages with pagesize 25)"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretPageSize/Paginate_100_secrets_with_pagesize_25 (44.38s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretPageSize/Paginate_30_secrets_with_pagesize_50
time="2025-07-09T12:54:43-07:00" level=info msg="Starting case: Paginate 30 secrets with pagesize 50"
time="2025-07-09T12:54:43-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:55:03-07:00" level=info msg="✓ Retrieved 30 secrets across 1 pages (expected 1 pages with pagesize 50)"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretPageSize/Paginate_30_secrets_with_pagesize_50 (20.87s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretPageSize/Paginate_73_secrets_with_pagesize_20
time="2025-07-09T12:55:04-07:00" level=info msg="Starting case: Paginate 73 secrets with pagesize 20"
time="2025-07-09T12:55:04-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:55:38-07:00" level=info msg="First page retrieved 20 items"
time="2025-07-09T12:55:38-07:00" level=info msg="✓ Retrieved 73 secrets across 4 pages (expected 4 pages with pagesize 20)"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretPageSize/Paginate_73_secrets_with_pagesize_20 (34.68s)
=== RUN   TestVaiTestSuite/TestVaiEnabled/SecretPageSize/Paginate_1_secret_with_pagesize_10
time="2025-07-09T12:55:38-07:00" level=info msg="Starting case: Paginate 1 secret with pagesize 10"
time="2025-07-09T12:55:38-07:00" level=info msg="Running with vai enabled: [true]"
time="2025-07-09T12:55:50-07:00" level=info msg="✓ Retrieved 1 secrets across 1 pages (expected 1 pages with pagesize 10)"
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretPageSize/Paginate_1_secret_with_pagesize_10 (11.41s)
--- PASS: TestVaiTestSuite/TestVaiEnabled/SecretPageSize (139.65s)
--- PASS: TestVaiTestSuite/TestVaiEnabled (311.68s)
--- PASS: TestVaiTestSuite (348.80s)
PASS
```

**✅ Expected Outcome**

For the automation to pass

**✅ Actual Outcome**

The automation passed

</details>
<hr>

# 2 / VAI Smarter Websockets Status: ✅ PASS

<a name="test-2"></a>

**:small_red_triangle: [back to top](#top)**

<details>
    <summary>Test 2 details... Click to expand</summary>

**Test Steps for Validation**

1. Run the following `script.sh` via `bash script.sh $RANCHER_URL $ADMIN_TOKEN`

<details>
    <summary>🧪 Test Script... CLICK TO EXPAND! ⬅️</summary>
<br>


```shell
#!/bin/bash

# Rancher SQLite Cache WebSocket Notification Testing
# Usage: ./test-websocket.sh <rancher-url> <admin-token>

set -e

if [ $# -ne 2 ]; then
    echo "Usage: $0 <rancher-url> <admin-token>"
    echo "Example: $0 https://rancher.example.com eyJhbGciOiJSUzI1NiIs..."
    exit 1
fi

RANCHER_URL="$1"
TOKEN="$2"
TEST_NS=""

echo "🔌 Rancher WebSocket Notification Test"
echo "======================================"
echo "Testing SQLite cache lightweight notifications"
echo "Rancher: $RANCHER_URL"
echo ""

setup_test_data() {
    echo "🚀 Setting up test data..."
    
    SUFFIX=$(openssl rand -hex 3)
    TEST_NS="ws-test-${SUFFIX}"
    
    kubectl create namespace "$TEST_NS"
    
    for i in {1..3}; do
        kubectl create secret generic "test-secret-$i" \
            --namespace="$TEST_NS" \
            --from-literal=data="value-$i" >/dev/null 2>&1
    done
    
    echo "✅ Created namespace: $TEST_NS"
}

show_websocket_setup() {
    echo ""
    echo "🔌 WebSocket Connection"
    echo "======================"
    echo ""
    echo "In a separate terminal, run:"
    echo ""
    echo "websocat -k \"wss://${RANCHER_URL#https://}/v1/subscribe\" \\"
    echo "  -H \"Authorization: Bearer $TOKEN\""
    echo ""
    echo "Then send this subscription (THE KEY TEST):"
    echo ""
    echo "{\"resourceType\":\"secrets\",\"namespace\":\"$TEST_NS\",\"mode\":\"resource.changes\"}"
    echo ""
    echo "Optional: Test with debouncing:"
    echo "{\"resourceType\":\"secrets\",\"namespace\":\"$TEST_NS\",\"mode\":\"resource.changes\",\"debounce\":1000}"
    echo ""
}

trigger_changes() {
    echo ""
    echo "⏸️  Set up WebSocket connection above, then press Enter to trigger changes..."
    read -r
    
    echo ""
    echo "🎯 Triggering changes (watch WebSocket for notifications)..."
    echo ""
    
    echo "🔥 Creating new secret..."
    kubectl create secret generic "new-secret" \
        --namespace="$TEST_NS" \
        --from-literal=test="websocket"
    echo "   Expected: {\"name\":\"resource.changed\",\"resourceType\":\"secrets\"}"
    sleep 2
    
    echo ""
    echo "🔥 Updating secret..."
    kubectl patch secret "test-secret-1" \
        --namespace="$TEST_NS" \
        -p '{"metadata":{"labels":{"updated":"true"}}}'
    echo "   Expected: {\"name\":\"resource.changed\",\"resourceType\":\"secrets\"}"
    sleep 2
    
    echo ""
    echo "🔥 Deleting secret..."
    kubectl delete secret "test-secret-2" --namespace="$TEST_NS"
    echo "   Expected: {\"name\":\"resource.changed\",\"resourceType\":\"secrets\"}"
    sleep 2
    
    echo ""
    echo "🔥 Rapid bulk changes (testing debouncing)..."
    for i in {4..6}; do
        kubectl create secret generic "bulk-$i" \
            --namespace="$TEST_NS" \
            --from-literal=bulk="true" &
    done
    wait
    echo "   Expected: Only 1-2 notifications due to debouncing"
    sleep 3
    
    echo ""
    echo "✅ All changes completed!"
}

show_expected_behavior() {
    echo ""
    echo "📋 Expected WebSocket Messages"
    echo "============================="
    echo ""
    echo "🟢 CORRECT (new SQLite cache behavior):"
    echo '{"name":"resource.changed","resourceType":"secrets"}'
    echo ""
    echo "🔴 INCORRECT (old behavior - if you see this, feature not working):"
    echo '{"name":"resource.change","resourceType":"secrets","object":{...full secret data...}}'
    echo ""
    echo "🎯 Success Criteria:"
    echo "   ✅ Small messages (~50 bytes, not 1KB+)"
    echo "   ✅ No 'object' field in messages"
    echo "   ✅ Debouncing: fewer notifications than changes"
    echo "   ✅ All operations (create/update/delete) trigger notifications"
}

cleanup() {
    echo ""
    echo "🧹 Cleanup"
    echo "=========="
    echo "Delete test namespace? (y/n)"
    read -r cleanup_choice
    if [ "$cleanup_choice" = "y" ]; then
        kubectl delete namespace "$TEST_NS" --ignore-not-found=true
        echo "✅ Deleted $TEST_NS"
    else
        echo "⚠️  Keeping $TEST_NS for further testing"
    fi
}

main() {
    setup_test_data
    show_websocket_setup
    trigger_changes
    show_expected_behavior
    cleanup
    
    echo ""
    echo "🎉 WebSocket testing complete!"
    echo ""
    echo "Summary: If you saw small notification messages instead of"
    echo "full resource objects, the SQLite cache WebSocket fix is working!"
}

main "$@"
```

</details>

Output:

```shell
websocat -k "wss://<REDACTED>/v1/subscribe" \
  -H "Authorization: Bearer token-xxxxx:<REDACTED>"
{"resourceType":"secrets","namespace":"ws-test-ec3626","mode":"resource.changes"}
{"name":"resource.start","namespace":"ws-test-ec3626","resourceType":"secrets","mode":"resource.changes","data":{"type":"subscribe","links":{}}}
{"name":"ping","data":{"type":"subscribe","links":{},"version":"dev"}}
{"name":"resource.changes","namespace":"ws-test-ec3626","resourceType":"secrets","mode":"resource.changes","data":{"type":"subscribe","links":{}}}
{"name":"resource.changes","namespace":"ws-test-ec3626","resourceType":"secrets","mode":"resource.changes","data":{"type":"subscribe","links":{}}}
{"name":"resource.changes","namespace":"ws-test-ec3626","resourceType":"secrets","mode":"resource.changes","data":{"type":"subscribe","links":{}}}
{"name":"ping","data":{"type":"subscribe","links":{},"version":"dev"}}
{"name":"ping","data":{"type":"subscribe","links":{},"version":"dev"}}
{"name":"resource.changes","namespace":"ws-test-ec3626","resourceType":"secrets","mode":"resource.changes","data":{"type":"subscribe","links":{}}}
```

**✅ Expected Outcome**

Minimal websocket responses

**✅ Actual Outcome**

Minimal websocket responses


</details>
<hr>

</details>
